src/test/app/MPToken_test.cpp

```diff
@@ -0,0 +1,1936 @@
+
+            MPTTester mptAlice(env, alice, {.holders = {carol}});
+
+            mptAlice.create({.ownerCount = 1, .holderCount = 0});
+
+            mptAlice.authorize({.account = carol});
+
+            // sendMax and DeliverMin are valid XRP amount,
+            // but is invalid combination with MPT amount
+            auto const MPT = mptAlice["MPT"];
+            env(pay(alice, carol, MPT(100)),
+                sendmax(XRP(100)),
+                ter(temMALFORMED));
+            env(pay(alice, carol, MPT(100)),
+                delivermin(XRP(100)),
+                ter(temBAD_AMOUNT));
+            // sendMax MPT is invalid with IOU or XRP
+            auto const USD = alice["USD"];
+            env(pay(alice, carol, USD(100)),
+                sendmax(MPT(100)),
+                ter(temMALFORMED));
+            env(pay(alice, carol, XRP(100)),
+                sendmax(MPT(100)),
+                ter(temMALFORMED));
+            env(pay(alice, carol, USD(100)),
+                delivermin(MPT(100)),
+                ter(temBAD_AMOUNT));
+            env(pay(alice, carol, XRP(100)),
+                delivermin(MPT(100)),
+                ter(temBAD_AMOUNT));
+            // sendmax and amount are different MPT issue
+            test::jtx::MPT const MPT1(
+                "MPT", makeMptID(env.seq(alice) + 10, alice));
+            env(pay(alice, carol, MPT1(100)),
+                sendmax(MPT(100)),
+                ter(temMALFORMED));
+            // paths is invalid
+            env(pay(alice, carol, MPT(100)), path(~USD), ter(temMALFORMED));
+        }
+
+        // build_path is invalid if MPT
+        {
+            Env env{*this, features};
+            Account const alice("alice");
+            Account const carol("carol");
+
+            MPTTester mptAlice(env, alice, {.holders = {bob, carol}});
+
+            mptAlice.create({.ownerCount = 1, .holderCount = 0});
+            auto const MPT = mptAlice["MPT"];
+
+            mptAlice.authorize({.account = carol});
+
+            Json::Value payment;
+            payment[jss::secret] = alice.name();
+            payment[jss::tx_json] = pay(alice, carol, MPT(100));
+
+            payment[jss::build_path] = true;
+            auto jrr = env.rpc("json", "submit", to_string(payment));
+            BEAST_EXPECT(jrr[jss::result][jss::error] == "invalidParams");
+            BEAST_EXPECT(
+                jrr[jss::result][jss::error_message] ==
+                "Field 'build_path' not allowed in this context.");
+        }
+
+        // Can't pay negative amount
+        {
+            Env env{*this, features};
+
+            MPTTester mptAlice(env, alice, {.holders = {bob}});
+
+            mptAlice.create({.ownerCount = 1, .holderCount = 0});
+            auto const MPT = mptAlice["MPT"];
+
+            mptAlice.authorize({.account = bob});
+
+            mptAlice.pay(alice, bob, -1, temBAD_AMOUNT);
+
+            env(pay(alice, bob, MPT(10)), sendmax(MPT(-1)), ter(temBAD_AMOUNT));
```

Comment:
Please add a few more cases here with `bob` trying to pay another holder (`carol`), and `bob` trying to pay back to `alice`. That'll probably require `alice` sending a valid amount to `bob` first.

Reply:
Do you mean paying negative amounts?

Reply:
Yeah, I think that's what I meant. I just wanted to make sure it didn't matter who was paying who for this test to fail.